src/EnergyPlus/HeatBalanceSurfaceManager.cc

```diff
@@ -6704,6 +6720,705 @@ namespace HeatBalanceSurfaceManager {
+
+                            TempSurfIn(surfNum) = TempSurfInTmp(surfNum);
+                        }
+                    }
+                }
+
+                // Loop over all HT surfaces
+                int const firstSurf = Zone(zoneNum).SurfaceFirst;
+                int const lastSurf = Zone(zoneNum).SurfaceLast;
+                for (int surfNum = firstSurf; surfNum <= lastSurf; ++surfNum) {
+                    auto &surface(Surface(surfNum));
+                    auto &zone(Zone(zoneNum));
+                    Real64 &TH11(TH(1, 1, surfNum));
+                    Real64 &TH12(TH(2, 1, surfNum));
+                    TH12 = TempSurfInRep(surfNum) = TempSurfIn(surfNum);
+                    TempSurfOut(surfNum) = TH11; // For reporting
+
+                    if (SurfaceWindow(surfNum).OriginalClass == SurfaceClass_TDD_Diffuser) { // Tubular daylighting device
+                        // Tubular daylighting devices are treated as one big object with an effective R value.
+                        // The outside face temperature of the TDD:DOME and the inside face temperature of the
+                        // TDD:DIFFUSER are calculated with the outside and inside heat balances respectively.
+                        // Below, the resulting temperatures are copied to the inside face of the TDD:DOME
+                        // and the outside face of the TDD:DIFFUSER for reporting.
+
+                        // Set inside temp variables of TDD:DOME equal to inside temp of TDD:DIFFUSER
+                        int domeNum = DataDaylightingDevices::TDDPipe(SurfaceWindow(surfNum).TDDPipeNum).Dome;
+                        TH(2, 1, domeNum) = TempSurfIn(domeNum) = TempSurfInTmp(domeNum) = TempSurfInRep(domeNum) = TempSurfIn(surfNum);
+
+                        // Set outside temp reporting variable of TDD:DOME (since it gets skipped otherwise)
+                        // Reset outside temp variables of TDD:DIFFUSER equal to outside temp of TDD:DOME
+                        TH11 = TempSurfOut(surfNum) = TempSurfOut(domeNum) = TH(1, 1, domeNum);
+                    }
+
+                    if ((TH12 > MaxSurfaceTempLimit) || (TH12 < MinSurfaceTempLimit)) {
+                        TestSurfTempCalcHeatBalanceInsideSurf(TH12, surface, zone, WarmupSurfTemp);
+                    }
+                }
+            } // ...end of main loops over all surfaces for inside heat balances
+
+            // Interzone surface updating: interzone surfaces have other side temperatures
+            // which can vary as the simulation iterates through the inside heat
+            // balance.  This block is intended to "lock" the opposite side (outside)
+            // temperatures to the correct value, namely the value calculated by the
+            // inside surface heat balance for the other side.
+            assert(TH.index(1, 1, 1) == 0u); // Assumed for linear indexing below
+            auto const l211(TH.index(2, 1, 1) - 1);
```

Comment:
Did not touch this loop. Refactoring TH will be fun - beware of stale comments which were not all updated when this 3-d array was flipped to row-major (which reordered the meaning of the 3 subscripts). @xuanluo113 @Myoldmopar